src/internal/packager2/remove.go

```
@@ -0,0 +1,160 @@
+						message.Warnf("Helm release for helm chart '%s' in the namespace '%s' was not found.  Was it already removed?", chart.ChartName, chart.Namespace)
+					}
+
+					// Pop the removed helm chart from the installed charts slice.
+					depPkg.DeployedComponents[len(depPkg.DeployedComponents)-i].InstalledCharts = depComp.InstalledCharts[:len(depComp.InstalledCharts)-1]
+					err = opt.Cluster.UpdateDeployedPackage(ctx, *depPkg)
+					if err != nil {
+						// We warn and ignore errors because we may have removed the cluster that this package was inside of
+						message.Warnf("Unable to update the secret for package %s, this may be normal if the cluster was removed: %s", depPkg.Name, err.Error())
+					}
+				}
+			}
+
+			err = actions.Run(ctx, comp.Actions.OnRemove.Defaults, comp.Actions.OnRemove.After, nil)
+			if err != nil {
+				return fmt.Errorf("unable to run the after action: %w", err)
+			}
+			err = actions.Run(ctx, comp.Actions.OnRemove.Defaults, comp.Actions.OnRemove.OnSuccess, nil)
+			if err != nil {
+				return fmt.Errorf("unable to run the success action: %w", err)
+			}
+
+			// Pop the removed component from deploy components slice.
+			if opt.Cluster != nil {
+				depPkg.DeployedComponents = depPkg.DeployedComponents[:len(depPkg.DeployedComponents)-1]
+				err = opt.Cluster.UpdateDeployedPackage(ctx, *depPkg)
+				if err != nil {
+					// We warn and ignore errors because we may have removed the cluster that this package was inside of
+					message.Warnf("Unable to update the secret for package %s, this may be normal if the cluster was removed: %s", depPkg.Name, err.Error())
+				}
+			}
+			return nil
+		}()
+		if err != nil {
+			err := actions.Run(ctx, comp.Actions.OnRemove.Defaults, comp.Actions.OnRemove.OnFailure, nil)
+			if err != nil {
+				// TODO: I think we should return an error if failure action fails. Does this break existing use cases?
+				return fmt.Errorf("unable to run the failure action: %w", err)
```

I want to start returning an error if on failure action fails. Before it was a log message. Does this break existing use cases?

---

It is a breaking change, but I would agree it makes sense to start erroring as we do with the other actions. We should make note of it in the release announcement.

---

This will affect people running Zarf in a pipeline as an exit code 1 would most likely stop the pipeline from continuing. I think it is worth causing some problems if it makes it clear to end users that an action actually failed instead of failing silently. I assume that most people are not reading their CI logs meticulously every time to catch the error log.